secret.go

```diff
@@ -3,27 +3,45 @@ package secretcrypt
 import (
 	"fmt"
 	"strings"
+	"sync"
 
 	"github.com/Zemanta/go-secretcrypt/internal"
 )
 
 // Secret represents an encrypted secret.
 type Secret struct {
+	lock          sync.RWMutex
 	crypter       internal.Crypter
 	ciphertext    internal.Ciphertext
+	plaintext     string
+	cached        bool
 	decryptParams internal.DecryptParams
 }
 
 // Decrypt decrypts the secret and returns the plaintext.
-func (s Secret) Decrypt() (string, error) {
+func (s *Secret) Decrypt() (string, error) {
 	if s.crypter == nil || s.ciphertext == "" {
 		return "", fmt.Errorf("Cannot decrypt a zero secret!")
 	}
-	return s.crypter.Decrypt(s.ciphertext, s.decryptParams)
+	s.lock.RLock()
```

Comment:
Since you check for cached in write lock later on, you don't need read lock around the first check.
It's true that with read lock you minimize the number of cases where you'll need write lock, but then you need read lock every time which probably affects average case much more than occasional unnecessary write lock.